test/sanity/code-smell/update-bundled.py

```diff
@@ -23,27 +23,46 @@
 
 import glob
 import json
+import os
 import os.path
-import sys
+import subprocess
 from distutils.version import LooseVersion
 
 import packaging.specifiers
 
 from ansible.module_utils.urls import open_url
 
 
-def compat_libs(basedir):
+def get_bundled_libs(basedir):
+    bundled_libs = set()
     for filename in glob.glob(os.path.join(basedir, 'lib/ansible/compat/*/__init__.py')):
         if 'compat/tests' in filename:
             # compat/tests doesn't bundle any code
             continue
-        yield filename
+        bundled_libs.add(filename)
 
-    yield os.path.join(basedir, 'lib/ansible/module_utils/distro/__init__.py')
-    yield os.path.join(basedir, 'lib/ansible/module_utils/six/__init__.py')
-    yield os.path.join(basedir, 'lib/ansible/module_utils/compat/ipaddress.py')
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/distro/__init__.py'))
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/six/__init__.py'))
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/compat/ipaddress.py'))
     # backports.ssl_match_hostname should be moved to its own file
-    yield os.path.join(basedir, 'lib/ansible/module_utils/urls.py')
+    bundled_libs.add(os.path.join(basedir, 'lib/ansible/module_utils/urls.py'))
+
+    return bundled_libs
+
+
+def get_files_with_bundled_metadata(basedir):
+    grep_cmd = subprocess.Popen(['grep', '-rwl', '--include=*.py', '_BUNDLED_METADATA', basedir],
```

Comment:
Use os.walk to find files and search them in Python instead of using grep.

Although ansible-test currently doesn't support passing a list of all files (regardless of changes) to code-smell tests, that is something that is going to be fixed. Once fixed, tests will use that list instead of os.walk to discover files. This will allow ansible-test to handle skipping files so that individual tests don't need to re-implement the logic.

If we use grep now then it will need to be replaced later once the file list is available. Using os.walk instead will make applying the changes much easier.

Reply:
I'd rather use grep as using a list of files (generated by os.walk or by ansible-test) will require reinventing smarts that grep already has builtin and since it does it in C, it's almost certainly faster as well.

Reply:
Using grep is almost certainly faster. However, if you're passing a list of files to grep you'll need to handle splitting up the file list into pieces small enough to avoid command line arg limits. Also, what grep "smarts" are you needing to use use?

Reply:
We don't have to pass a list of files to grep if we're operating on all files, right?  We just need to pass it the top of the tree.

In this case the smarts are word boundaries and include by file extension.  These are doable in python but I'm questioning what we gain by pushing the implementation of these things into our python code instead of letting grep handle those details?  It's not complexity and it's not speed but if there's another reason, then that's fine, just let me know what it is...

Reply:
I've fixed the limitation in ansible-test that prevented code-smell tests from receiving file lists regardless of changed files. This test can now use the ansible-test provided file list and will not need to use os.walk or grep.

The `update-bundled.json` file should be changed as follows:

```json
{
    "ignore_changes": true,
    "extensions": [
        ".py"
    ],
    "output": "path-message"
}
```

The script only needs the following to get the list of files:

```python
paths = sys.argv[1:] or sys.stdin.read().splitlines()
```

As for the word boundary handling, a Python regex deals with that nicely:

```python
for path in paths:
    with open(path, 'r') as path_fd:
        contents = path_fd.read()

    if re.search(r'\b_BUNDLED_METADATA\b', contents):
        pass  # do something here
```

Use of the file list provided by ansible-test means that the script does not need to know anything about files or directories which should be skipped, as that logic already exists in ansible-test. For example, directories like .git, .tox, etc. are skipped. As the logic is changed/improved in ansible-test all of the code-smell tests benefit automatically. In many cases this can lead to improved performance since the test is checking far fewer files, even though the checks themselves may be a bit slower since they're in python.

Another benefit of avoiding grep is that we don't need to be worried about avoiding options that are not available in all versions of grep, since the logic is all in the python code.

If the test only needs to look at certain directories, ansible-test will filter those out as well. Here's an example configuration:

https://github.com/ansible/ansible/blob/06d83bae056bbc9de904a238e480acf1147fe80e/test/sanity/code-smell/no-main-display.json#L5-L8

Reply:
<nod>  Yep, a Python regex can do what grep does but it is still taking the things that grep has been optimized to do and attempting to write our own code that can only do it slower and with more complexity than shelling out.  But if portability is the overriding concern then I can make that change.